
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>statSTR package &#8212; STRTools 0.0.1 documentation</title>
    <link rel="stylesheet" href="../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/language_data.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="statstr-package">
<h1>statSTR package<a class="headerlink" href="#statstr-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-statSTR.statSTR">
<span id="statstr-statstr-module"></span><h2>statSTR.statSTR module<a class="headerlink" href="#module-statSTR.statSTR" title="Permalink to this headline">¶</a></h2>
<p>Tool for computing stats on STR VCF files</p>
<dl class="function">
<dt id="statSTR.statSTR.GetAFreq">
<code class="sig-prename descclassname">statSTR.statSTR.</code><code class="sig-name descname">GetAFreq</code><span class="sig-paren">(</span><em class="sig-param">trrecord</em>, <em class="sig-param">samplelists=[]</em>, <em class="sig-param">count=False</em>, <em class="sig-param">uselength=True</em><span class="sig-paren">)</span><a class="headerlink" href="#statSTR.statSTR.GetAFreq" title="Permalink to this definition">¶</a></dt>
<dd><p>Return allele frequency for a TR</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>trrecord</strong> (<em>trh.TRRecord object</em>) – The record that we are computing the statistic for</p></li>
<li><p><strong>samplelist</strong> (<em>list of list of str</em>) – List of lists of the samples that we include when compute the statistic</p></li>
<li><p><strong>count</strong> (<em>bool</em>) – If True, return allele counts rather than allele frequencies</p></li>
<li><p><strong>uselength</strong> (<em>bool</em>) – Whether we should collapse alleles by length</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>allele_freqs_strs</strong> – Format: allele1:freq1,allele2:freq2,etc. for each sample group</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list of str</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="statSTR.statSTR.GetHWEP">
<code class="sig-prename descclassname">statSTR.statSTR.</code><code class="sig-name descname">GetHWEP</code><span class="sig-paren">(</span><em class="sig-param">trrecord</em>, <em class="sig-param">samplelists=[]</em>, <em class="sig-param">uselength=True</em><span class="sig-paren">)</span><a class="headerlink" href="#statSTR.statSTR.GetHWEP" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute Hardy Weinberg p-value</p>
<p>Tests whether the number of observed heterozygous vs.
homozygous individuals is different than expected
under Hardy Weinberg Equilibrium given the observed
allele frequencies, based on a binomial test.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>trrecord</strong> (<em>trh.TRRecord object</em>) – The record that we are computing the statistic for</p></li>
<li><p><strong>samplelist</strong> (<em>list of list of str</em>) – List of list of the samples that we include when compute the statistic</p></li>
<li><p><strong>uselength</strong> (<em>bool</em>) – Whether we should collapse alleles by length</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>p-value</strong> – The two-sided p-value returned by a binomial test (scipy.stats.binom_test)
If the allele frequencies dictionary is invalid, return np.nan
If the genotype alleles not included in frequencies dictionary, return np.nan
One value returned for each samplelist</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list of float</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="statSTR.statSTR.GetHeader">
<code class="sig-prename descclassname">statSTR.statSTR.</code><code class="sig-name descname">GetHeader</code><span class="sig-paren">(</span><em class="sig-param">header</em>, <em class="sig-param">sample_prefixes</em><span class="sig-paren">)</span><a class="headerlink" href="#statSTR.statSTR.GetHeader" title="Permalink to this definition">¶</a></dt>
<dd><p>Return header items for a column</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>header</strong> (<em>str</em>) – Header item</p></li>
<li><p><strong>sample_prefixes</strong> (<em>list of str</em>) – List of sample prefixes. empty if no sample groups used</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>header_items</strong> – List of header items</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list of str</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="statSTR.statSTR.GetHet">
<code class="sig-prename descclassname">statSTR.statSTR.</code><code class="sig-name descname">GetHet</code><span class="sig-paren">(</span><em class="sig-param">trrecord</em>, <em class="sig-param">samplelists=[]</em>, <em class="sig-param">uselength=True</em><span class="sig-paren">)</span><a class="headerlink" href="#statSTR.statSTR.GetHet" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute heterozygosity of a locus</p>
<p>Heterozygosity is defined as the probability
that two randomly drawn allele are different.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>trrecord</strong> (<em>trh.TRRecord object</em>) – The record that we are computing the statistic for</p></li>
<li><p><strong>samplelist</strong> (<em>list of list of str</em>) – List of list of the samples that we include when compute the statistic</p></li>
<li><p><strong>uselength</strong> (<em>bool</em>) – Whether we should collapse alleles by length</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>heterozygosity</strong> – The heterozygosity of the locus. One value for each sample list.
If the allele frequencies dictionary is invalid, return np.nan</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list of float</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="statSTR.statSTR.GetMean">
<code class="sig-prename descclassname">statSTR.statSTR.</code><code class="sig-name descname">GetMean</code><span class="sig-paren">(</span><em class="sig-param">trrecord</em>, <em class="sig-param">samplelists=[]</em>, <em class="sig-param">uselength=True</em><span class="sig-paren">)</span><a class="headerlink" href="#statSTR.statSTR.GetMean" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute the mean allele length</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>trrecord</strong> (<em>trh.TRRecord object</em>) – The record that we are computing the statistic for</p></li>
<li><p><strong>samplelist</strong> (<em>list of list of str</em>) – List of list of the samples that we include when compute the statistic</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>mean</strong> – The mean allele length. One value for each sample list
If the allele frequencies dictionary is invalid, return np.nan</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list of float</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="statSTR.statSTR.GetMode">
<code class="sig-prename descclassname">statSTR.statSTR.</code><code class="sig-name descname">GetMode</code><span class="sig-paren">(</span><em class="sig-param">trrecord</em>, <em class="sig-param">samplelists=[]</em><span class="sig-paren">)</span><a class="headerlink" href="#statSTR.statSTR.GetMode" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute the mode of the allele frequencies</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>trrecord</strong> (<em>trh.TRRecord object</em>) – The record that we are computing the statistic for</p></li>
<li><p><strong>samplelist</strong> (<em>list of list of str</em>) – List of the samples that we include when compute the statistic</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>mode</strong> – The mode of the allele frequencies. One value for each sample list
If the allele frequencies dictionary is invalid, return np.nan</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list of float</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="statSTR.statSTR.GetNumSamples">
<code class="sig-prename descclassname">statSTR.statSTR.</code><code class="sig-name descname">GetNumSamples</code><span class="sig-paren">(</span><em class="sig-param">trrecord</em>, <em class="sig-param">samplelists=[]</em><span class="sig-paren">)</span><a class="headerlink" href="#statSTR.statSTR.GetNumSamples" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute the number of samples</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>trrecord</strong> (<em>trh.TRRecord object</em>) – The record that we are computing the statistic for</p></li>
<li><p><strong>samplelist</strong> (<em>list of list of str</em>) – List of list of the samples that we include when compute the statistic</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>numSamples</strong> – The number of samples. One value for each sample list
If the allele frequencies dictionary is invalid, return np.nan</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list of int</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="statSTR.statSTR.GetThresh">
<code class="sig-prename descclassname">statSTR.statSTR.</code><code class="sig-name descname">GetThresh</code><span class="sig-paren">(</span><em class="sig-param">trrecord</em>, <em class="sig-param">samplelists=[]</em><span class="sig-paren">)</span><a class="headerlink" href="#statSTR.statSTR.GetThresh" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the maximum TR allele length observed</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>trrecord</strong> (<em>trh.TRRecord object</em>) – The record that we are computing the statistic for</p></li>
<li><p><strong>samplelists</strong> (<em>list of list of str</em>) – List of lists of the samples that we include when compute the statistic</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>thresh</strong> – List of Maximum allele length observed in each sample group</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list of float</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="statSTR.statSTR.GetVariance">
<code class="sig-prename descclassname">statSTR.statSTR.</code><code class="sig-name descname">GetVariance</code><span class="sig-paren">(</span><em class="sig-param">trrecord</em>, <em class="sig-param">samplelists=[]</em><span class="sig-paren">)</span><a class="headerlink" href="#statSTR.statSTR.GetVariance" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute the variance of the allele lengths</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>trrecord</strong> (<em>trh.TRRecord object</em>) – The record that we are computing the statistic for</p></li>
<li><p><strong>samplelists</strong> (<em>list of list of str</em>) – List of list of the samples that we include when compute the statistic</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>variance</strong> – The variance of the allele lengths. One value for each sample list
If the allele frequencies dictionary is invalid, return np.nan</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list of float</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="statSTR.statSTR.PlotAlleleFreqs">
<code class="sig-prename descclassname">statSTR.statSTR.</code><code class="sig-name descname">PlotAlleleFreqs</code><span class="sig-paren">(</span><em class="sig-param">trrecord</em>, <em class="sig-param">outprefix</em>, <em class="sig-param">samplelists=[]</em>, <em class="sig-param">sampleprefixes=[]</em><span class="sig-paren">)</span><a class="headerlink" href="#statSTR.statSTR.PlotAlleleFreqs" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot allele frequencies for a locus</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>trrecord</strong> (<em>trh.TRRecord object</em>) – The record that we are computing the statistic for</p></li>
<li><p><strong>outprefix</strong> (<em>str</em>) – Prefix for output file</p></li>
<li><p><strong>samplelists</strong> (<em>list of list of str</em><em>, </em><em>optional</em>) – List of lists of the samples that we include when compute the statistic</p></li>
<li><p><strong>sampleprefixes</strong> (<em>list of str</em><em>, </em><em>optional</em>) – Prefixes for each sample list to use in legend</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="statSTR.statSTR.getargs">
<code class="sig-prename descclassname">statSTR.statSTR.</code><code class="sig-name descname">getargs</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#statSTR.statSTR.getargs" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="statSTR.statSTR.main">
<code class="sig-prename descclassname">statSTR.statSTR.</code><code class="sig-name descname">main</code><span class="sig-paren">(</span><em class="sig-param">args</em><span class="sig-paren">)</span><a class="headerlink" href="#statSTR.statSTR.main" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</div>
<div class="section" id="statstr-test-statstr-module">
<h2>statSTR.test_statSTR module<a class="headerlink" href="#statstr-test-statstr-module" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-statSTR">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-statSTR" title="Permalink to this headline">¶</a></h2>
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">STRTools</a></h1>








<h3>Navigation</h3>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2020, Gymreklab.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 2.3.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="../_sources/source/statSTR.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>