.. overview_directive
.. |prancSTR overview| replace:: prancSTR identifies STRs with evidence of somatic mosaicism. prancSTR only works with HipSTR VCFs.
.. overview_directive_done

prancSTR
=========

Population-level heterogeneity arises due to germline mutations that occur before the formation of the zygote and are inherited by all cells in the offspring.
However, heterogeneity within an individual may also exist due to somatic mutations that occur post-zygotically in only a sub-population of cells.

prancSTR is a tool for detecting somatic mosaicism at STRs using HipSTR calls. Note that prancSTR does not require a matched control sample as input, and currently only works with HipSTR input. 

prancSTR uses the following fields from HipSTR's output for detecting mosaicism:
* estimated diploid repeat lengths
* the observed distribution of copy numbers across all reads aligning to a locus
* diploid genotype
* stutter params
* observed allele length distributions (MALLREADS)

Somatic mosaicism at STRs is known to play a key role in the pathogenicity of loci implicated in repeat expansion disorders and non-neoplastic syndromes.

prancSTR is in *beta*.

Usage
-----
To run prancSTR use the following command::

	prancSTR \
  	  --vcf <vcf file> \
  	  --out <string> \
  	  [filter options]

Required parameters:

* :code:`--vcf <VCF>` Input VCF file, generated by HipSTR. 
* :code:`--out <string>` Output file prefix. Use :code:`stdout` to print file to standard output

prancSTR will output a tab-delimited file with predicted mosaicism at STR loci either to stdout or to :code:`$out.tab`. See a description of the output file below.

Other general parameters:

* :code:`--region <string>`: Restrict to the region chr:start-end. VCF file must be bgzipped and indexed to use this option.
* :code:`--samples <string>`: Restrict to the given list of samples. Samples are comma separated.
* :code:`--vcftype <string>`: Specify the tool which generated the vcf call file for STRs. e.g. hipstr, gangstr etc.
* :code:`--only-passing <action=store_true>`: Filters out the regions which don't have the FILTER column set to PASS.
* :code:`--output-all <action=store_true>`: Force tool to output results for all loci.
* :code:`--readfield <string>`: Specify which field to utilize for extracting read information.
* :code:`--debug <action=store_true>`: Print helpful debug messages.
* :code:`--quiet <action=store_true>`: Restrict printing of any messages.
* :code:`--version`: Print the version of the tool

Note: FILTER column is generated by dumpSTR or other filtering tools based on the set of parameters users want to filter out their TR calls on. 
FILTER column therefore only serves to reflect whether any given locus passes the set filtering criteria or not.

See `Example Commands`_ for examples running prancSTR under different settings.


Output files
------------

prancSTR output file contains mosaicism predictions generated for each locus. It is up to the user's discretion to filter out for high confidence mosaic allele calls.
The output generated is a tab-delimited file with 16 columns where each row represents predicted mosaicism at STR:

* :code:`sample`: The ID of the sample whose locus is being analyzed.
* :code:`chrom`: The chromosome on which the mosaic short tandem repeat (STR) lies.
* :code:`locus`: Reference ID for the short tandem repeat.
* :code:`motif`: The nucleotide sequence that the repeat is comprised of.
* :code:`A`: Represents the first genotype allele for the given STR in repeat units relative to the reference.
* :code:`B`: Represents the second genotype allele for the given STR in repeat units relative to the reference.
* :code:`C`: This is the mosaic allele inferred by prancSTR in repeat units relative to the reference.
* :code:`f`: This represents the mosaic allele fraction. 
* :code:`pval`: Gives the p-value metric for how significant our findings are.
* :code:`reads`: Gives representation for how many reads support each allele.
* :code:`mosaic_support`: The number of reads that support the mosaic allele. 
* :code:`stutter parameter u`: Gives the probability that stutter error causes an increase in obs. STR size.
* :code:`stutter parameter d`: Gives the probability that stutter error causes a decrease in obs. STR size.
* :code:`stutter parameter rho`: Parameter for geometric step size distribution.
* :code:`quality factor`: Reports the posterior probability of the genotype as a mesaure of quality of individual sample's genotype.
* :code:`read depth`: Reports the total depth/number of informative reads for all samples at the locus.

Stutter parameter u, stutter parameter d, stutter parameter rho, quality factor and read depth are the parameters that have been imported from what HipSTR outputs.
Below is an example file which contains 5 STR loci 

+---------+-------+---------+---------------+-------+----+---+---+----------+--------------+------------------+----------------+---------------------+--------------------+----------------------+----------------+------------+
| sample  | chrom |   pos   |     locus     | motif | A  | B | C |    f     |     pval     |      reads       | mosaic_support | stutter parameter u | stutter paramter d | stutter paramter rho | quality factor | read depth |
+=========+=======+=========+===============+=======+====+===+===+==========+==============+==================+================+=====================+====================+======================+================+============+
| NA07022 | chr1  | 987287  | Human_STR_285 |   T   | 3  | 5 | 2 | 0.244079 | 1.530865e-04 | 2|4;3|4;5|4;6|1  |       4        |        0.01         |        0.07        |         0.31         |      0.98      |     21     |
+---------+-------+---------+---------------+-------+----+---+---+----------+--------------+------------------+----------------+---------------------+--------------------+----------------------+----------------+------------+
| NA12716 | chr1  | 987287  | Human_STR_285 |   T   | 5  | 5 | 4 | 0.265958 | 1.842418e-05 |   4|6;5|15;9|1   |       6        |        0.01         |        0.07        |         0.31         |      1.00      |     34     |
+---------+-------+---------+---------------+-------+----+---+---+----------+--------------+------------------+----------------+---------------------+--------------------+----------------------+----------------+------------+
| NA06989 | chr1  | 1002414 | Human_STR_295 |   T   | -1 | 4 | 3 | 0.150960 | 6.045544e-05 |  -1|16;3|4;4|9   |       4        |        0.02         |        0.02        |         0.69         |      1.00      |     50     |
+---------+-------+---------+---------------+-------+----+---+---+----------+--------------+------------------+----------------+---------------------+--------------------+----------------------+----------------+------------+
| NA10847 | chr1  | 1002414 | Human_STR_295 |   T   | 4  | 5 | 3 | 0.280689 | 1.290112e-09 | 2|1;3|7;4|6;5|11 |       7        |        0.02         |        0.02        |         0.69         |      1.00      |     55     |
+---------+-------+---------+---------------+-------+----+---+---+----------+--------------+------------------+----------------+---------------------+--------------------+----------------------+----------------+------------+
| NA12347 | chr1  | 1002414 | Human_STR_295 |   T   | 5  | 5 | 4 | 0.262358 | 1.029537e-05 | 3|1;4|5;5|14;6|1 |       5        |        0.02         |        0.02        |         0.69         |      0.99      |     51     |
+---------+-------+---------+---------------+-------+----+---+---+----------+--------------+------------------+----------------+---------------------+--------------------+----------------------+----------------+------------+

We filter output on the following parameters to obtain what we consider to be high confidence mosaic sites:

* :code:`pval`:of less than equal to 0.05
* :code:`read depth`:of greater than equal to 10
* :code:`quality factor` of greater than equal to 0.8
* :code:`mosaic_support` of greater than equal to 3
* :code:`f`: of less than equal to 0.3 or 30%


Example Commands
----------------

Below are :code:`prancSTR` examples using HipSTR VCFs. Data files can be found at https://github.com/gymreklab/TRTools/tree/master/example-files::

	# Example command running prancSTR for only one chromosome with hipstr output file while filtering segemental duplicates
	prancSTR \
	   --vcf CEU_subset.vcf.gz \
	   --out CEU_chr1  \
	   --vcftype hipstr \
	   --only-passing \
	   --region chr1

	# Example command running prancSTR for only one sample while filtering segemental duplicates
	prancSTR \
	   --vcf CEU_subset.vcf.gz \
	   --only-passing \
	   --out NA12878_mosaicSTR \
	   --samples NA12878


Citations
----------------
If you use prancSTR in your work, please cite: 