# .. overview_directive
.. |annotaTR overview| replace:: AnnotaTR takes in a TR genotype VCF file and outputs an new file (VCF or other formats) with additional INFO/FORMAT fields
.. overview_directive_done


annotaTR
========

AnnotaTR can be used to add information to a TR genotype file. It can output annotated files in either VCF or PGEN format. Current use cases include:

* Annotating a TR VCF file with Beagle-imputed genotypes to add back TR metadata from a reference panel
* Computing dosages for each genotype, which can be used for downstream association testing.

A more detailed list of use cases is provided below.

|annotaTR overview|

Usage
-----
The following is a basic annotaTR command::

	annotaTR \
  	  --vcf <vcf file> \
  	  --out <string> \
  	  [annotation options]

Required parameters:

* :code:`--vcf <VCF>`: VCF file to filter that has been generated by a supported genotyping tool (or imputed from one using Beagle)
* :code:`--out <string>`: prefix to name output files.

Other general parameters:

* :code:`--vcftype <string>`: Which genotyping tool generated the input VCF. Default = :code:`auto`. Necessary if it cannot be automatically inferred. One of: :code:`gangstr`, :code:`advntr`, :code:`hipstr`, :code:`eh`.
* :code:`--outtype <string>`: Which output format to generate. Supported arguments are :code:`vcf` or :code:`pgen`. If a comma-separated listed of types is specified (e.g. :code:`vcf,pgen`), all specified output formats are generated.

In addition to specifying input and output options above, you must specify at least one annotation operation to perform. These are described below.

Annotation options
------------------

annotaTR offers several annotation options:

Computing dosages
^^^^^^^^^^^^^^^^^

Dosages are quantitative representations of individual-level genotypes. Use cases and advantages/disadvantages of this representation are described below. The following option computes dosages:

:code:`--dosages <string>`: The string argument to this option specifies the method to compute dosages. It must be one of: 

* :code:`bestguess`: dosages are computed by summing the length of alleles (given in number of repeat units) for each call. e.g. for a genotype heterozygous for 3 and 4 copies of a repeat, the best guess dosage will be 7.
* :code:`bestguess_norm`: same as above, but scaled to be between 0 and 2. This is required if generating pgen output, since pgen only supports dosage values in this range.
* :code:`beagpleap`: dosages for imputed calls are computed in a way that takes into account imputation uncertainty. This option requires the Beagle VCF FORMAT fields :code:`AP1` and :code:`AP2` to be present, which give the probability of each alternate allele on each of the two haplotypes of an individual. AP-based dosages are generated by computing a weighted sum of allele lengths for each haplotype (sum of alleleprob*allelelength over all alleles) and added together to get the total dosage for the genotype. In the case of imputation with no uncertainty (all AP fields are either 0 or 1), dosages computed in this way should match best guess dosages.
* :code:`beagleap_norm`: same as above, but scaled to be between 0 and 2.

If annotating dosages, the following fields are added to VCF output files:

* FORMAT field :code:`TRDS`. This is a float value giving the estimated TR dosage for each call.
* INFO field :code:`DSLEN`. This records the minimum and maximum allele lengths, which can be useful to know if computing normalized dosages but you later want to recover the unnormalized dosage values.

Dosages may also be output to PGEN format. Because dosages are not explicitly supported for multi-allelic sites, we output an accompanying dummy pvar file in which:

* All TR variants are listed with reference alleles of "A" and alternate alleles of "T". These are just placeholders since those fields are required.
* We include the DSLEN field described above under VCF format.

Annotating imputed TR VCFs
^^^^^^^^^^^^^^^^^^^^^^^^^^
TODO

Notes on output files
^^^^^^^^^^^^^^^^^^^^^
TODO

Use cases
^^^^^^^^^

# TODO below under construction
A major use case for dosages is to perform association testing between TR genotypes and a phenotype of interest. Major advantages of using dosages are:

* They can be used as input to downstream tools that support association testing with dosage values, even if those tools do not explicitly support TRs or multi-allelic sites. Below we discuss how to use annotaTR-generated pgen files with dosages as input to plink.
* This enables testing TRs alongside other variant types using the same pipelines, streamlining workflows that include multiple variant types.

